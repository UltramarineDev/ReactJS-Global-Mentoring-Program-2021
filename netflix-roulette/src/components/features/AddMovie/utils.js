export const getInitialValues = (movie = {}) => ({
  title: movie.title || '',
  releaseDate: movie.release_date || '',
  movieUrl: movie.poster_path || '',
  genre: movie.genres ? movie.genres[0] : '',
  overview: movie.overview || '',
  runtime: movie.runtime || '',
});

export const buildAddMovieInput = (values) => {
  const { title, releaseDate, movieUrl, genre, overview, runtime } = values;
  return {
    title,
    release_date: releaseDate,
    poster_path: movieUrl,
    genres: [genre],
    overview,
    runtime: runtime || 0,
  };
};
