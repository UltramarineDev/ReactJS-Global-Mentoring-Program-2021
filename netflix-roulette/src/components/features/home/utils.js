import { TABS, SORT_OPTIONS, SORT_OPTIONS_LABELS, FILTER_TAGS } from './constants';

export const getTabs = () => ([
  {
    label: 'ALL',
    id: TABS.ALL,
  },
  {
    label: 'DOCUMENTARY',
    id: TABS.DOCUMENTARY,
  },
  {
    label: 'COMEDY',
    id: TABS.COMEDY,
  },
  {
    label: 'HORROR',
    id: TABS.HORROR,
  },
  {
    label: 'CRIME',
    id: TABS.CRIME,
  },
]);

export const getFilteredMovies = (movies, activeTab) => {
  const filterTagList = FILTER_TAGS[activeTab];
  
  if (!movies || movies.length === 0 || !filterTagList) {
    return movies;
  }

  let filteredMovies = movies.filter(({ genres }) => {
    if(!genres || genres.length === 0) {
      return false;
    }

    let matchFound = false;

    genres.forEach((genre) => {
      const areTagsContainsGenre = filterTagList.some((filterTag) => filterTag === genre);
      if (areTagsContainsGenre) {
        matchFound = true;
        return matchFound;
      }
    });

    return matchFound;
  });

  return filteredMovies;
};

export const getOptions = (options, labels) => {
  const resultIds = Object.values(options);

  return resultIds.map(resultId => ({
    id: resultId,
    name: labels[resultId]
  }))
};

//TODO: implement ui sorting in scope of module 4
export const getSortedMovies = (movies, value) => {
  return movies;
};
