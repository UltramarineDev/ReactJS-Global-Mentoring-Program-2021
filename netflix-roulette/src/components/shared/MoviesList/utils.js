import { tabs, filterTags } from './constants';
import { wordings } from '/src/locales/wordings';

export const getTabs = () => ([
  {
    label: wordings.all,
    id: tabs.ALL,
  },
  {
    label: wordings.documentary,
    id: tabs.DOCUMENTARY,
  },
  {
    label: wordings.comedy,
    id: tabs.COMEDY,
  },
  {
    label: wordings.horror,
    id: tabs.HORROR,
  },
  {
    label: wordings.crime,
    id: tabs.CRIME,
  },
]);

export const getOptions = (options, labels) => {
  const resultIds = Object.values(options);

  return resultIds.map(resultId => ({
    id: resultId,
    name: labels[resultId]
  }))
};

//TODO: implement ui sorting in scope of module 4
export const getSortedMovies = (movies, value) => {
  return movies;
};
