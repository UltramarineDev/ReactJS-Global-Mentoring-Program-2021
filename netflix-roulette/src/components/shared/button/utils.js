import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
//TODO: update BUTTON_ICONS
import { BUTTON_ICONS } from '/src/components/shared/constants';

export const getButtonIcon = (icon) => {
  switch (icon) {
    case BUTTON_ICONS.ADD:
      return <FontAwesomeIcon icon="plus" />;
    default: 
      return undefined;
  }
};
