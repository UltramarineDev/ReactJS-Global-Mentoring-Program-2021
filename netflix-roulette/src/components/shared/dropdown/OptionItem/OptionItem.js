import React from 'react';
import PropTypes from 'prop-types';

import styles from './OptionItem.module.scss';

const OptionItem = ({ label, onClick, value }) => {
  const onLinkClick = () => onClick(value);

  return (
    <a aria-label="option_item" onClick={onLinkClick} className={styles.item}>{label}</a>
  );
};

OptionItem.propTypes = {
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  value: PropTypes.object.isRequired,
};

export default OptionItem;
