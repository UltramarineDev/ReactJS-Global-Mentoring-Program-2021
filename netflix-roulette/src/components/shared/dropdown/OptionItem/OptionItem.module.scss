.item {
  display: block;
  padding: 12px 12px;
}

.item:hover {
  background-color: #f65261;
}
