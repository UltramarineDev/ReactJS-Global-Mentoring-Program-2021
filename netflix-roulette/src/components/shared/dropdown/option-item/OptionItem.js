import React from 'react';
import PropTypes from 'prop-types';

import styles from './OptionItem.module.scss';

const OptionItemComponent = ({ label, onClick }) => (
  <>
    <div className={styles.item}>
      <a onClick={onClick}>{label}</a>
    </div>
  </>
);

OptionItemComponent.propTypes = {
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default OptionItemComponent;
