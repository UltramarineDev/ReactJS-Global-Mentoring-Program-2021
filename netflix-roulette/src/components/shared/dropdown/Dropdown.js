import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classnames from 'classnames';

import OptionItem from './OptionItem/OptionItem';
import styles from './Dropdown.module.scss';

const Dropdown = ({
  value,
  options,
  onChange,
  placeholder,
  className,
}) => {
  const currentValueLabel = value ? options.find(({ id }) => id === value).name : placeholder;
  const [showOptions, setShowOptions] = useState(false);

  const onChangeClick = useCallback((value) => {
    setShowOptions((showOptions) => !showOptions);
    onChange(value);
  }, []);

  return (
    <div className={styles.dropdown}>
      <a onClick={() => setShowOptions((showOptions) => !showOptions)}>{currentValueLabel}</a>
      {showOptions
        ? <div className={styles.icon}><FontAwesomeIcon icon="sort-up" /></div>
        : <div className={styles.icon}><FontAwesomeIcon icon="sort-down" /></div>}
      {showOptions
        && (
        <div className={classnames(styles.menu, className)}>
          {options.map(({ id, name }) => (
            <OptionItem
              key={id}
              value={id}
              label={name}
              onClick={onChangeClick}
            />
          ))}
        </div>
        )}
    </div>
  );
};

Dropdown.propTypes = {
  value: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  })).isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  className: PropTypes.string,
};

Dropdown.defaultProps = {
  placeholder: '',
  className: '',
};

export default Dropdown;
