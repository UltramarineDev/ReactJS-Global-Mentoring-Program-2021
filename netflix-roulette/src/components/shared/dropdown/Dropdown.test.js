import React from 'react';
import { fireEvent, screen, render, queryByAttribute, cleanup } from '@testing-library/react';

import { getGenres } from 'components/features/MovieForm/utils';

import Dropdown from './Dropdown';
import OptionItem from './OptionItem/OptionItem';

describe('Dropdown tests', () => {
  beforeEach(() => {
    jest.resetModules();
  });
  afterEach(cleanup);

  test('renders Dropdown', () => {
    const onChange = jest.fn();
    const value = '';
    const options = getGenres();
    const { asFragment } = render(<Dropdown value={value} options={options} onChange={onChange} />);
    expect(asFragment(<Dropdown value={value} options={options} onChange={onChange} />)).toMatchSnapshot();
  });

  test('renders Dropdown with showOptions', () => {
    const onChange = jest.fn();
    const value = '';
    const options = getGenres();
    const dom = render(<Dropdown value={value} options={options} onChange={onChange} />);

    const getById = queryByAttribute.bind(null, 'id');
    const link = getById(dom.container, 'set_options_dropdown');

    fireEvent.click(link);
    expect(screen.getAllByLabelText('option_item')).toHaveLength(5);
  });

  test('Dropdown with onChangeClick', () => {
    const onChange = jest.fn();
    const value = '';
    const options = getGenres();
    const { getByLabelText, getByText } = render(<Dropdown value={value} options={options} onChange={onChange} />);

    const link = getByLabelText('set_options_dropdown_label');
    fireEvent.click(link);

    const optionItem = getByText('COMEDY');
    fireEvent.click(optionItem);

    expect(onChange).toBeCalled();
  });
});
