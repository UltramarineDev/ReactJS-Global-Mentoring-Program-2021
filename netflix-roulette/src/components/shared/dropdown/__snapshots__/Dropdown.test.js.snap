// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown tests renders Dropdown 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <a
      aria-label="set_options_dropdown_label"
      id="set_options_dropdown"
    />
    <div
      class="icon"
    />
  </div>
</DocumentFragment>
`;
